Sock ctrl msg

uran0sH opened the pull request.

uran0sH committed:
Use pipefd to replace eventfd in tests

Eventfd is linux-specific. In order to support macos
and others, we need to replace it with pipefd.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

uran0sH committed:
Replace CMSG_*! macros with libc equivalent

libc also provides functions with the same functionality
and supports multiple platforms, so we can replace
CMSG_*! with libc equivalent.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

uran0sH committed:
Move sock_ctrl_msg.rs to the unix folder

Make some modifications to sock_ctl_msg to support Unix systems.
Including adjusting the size of cmsg_len for different platforms.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

uran0sH commented:
Plz review @stefano-garzarella @germag @osteffenrh

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -555,16 +504,18 @@ mod tests {
         let (s1, s2) = UnixDatagram::pair().expect("failed to create socket pair");
 
-        let evt = EventFd::new(0).expect("failed to create eventfd");
+        let evt = pipe().expect("failed to create pipefd");
```
I already mentioned to keep all the EventFd -> pipe() changes in the same commit, same below.
Is something that I'm missing that requires this changes to be in this commit?

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -602,18 +555,18 @@ mod tests {
+                    evt1.1.as_raw_fd(),
+                    evt2.1.as_raw_fd(),
+                    evt3.1.as_raw_fd(),
+                    evt4.1.as_raw_fd(),
```
this changes should belong to the first commit

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -635,18 +588,18 @@ mod tests {
+                    evt1.1.as_raw_fd(),
+                    evt2.1.as_raw_fd(),
+                    evt3.1.as_raw_fd(),
+                    evt4.1.as_raw_fd(),
```
move to first commit

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -586,14 +537,16 @@ mod tests {
 
-        assert_eq!(evt.read().expect("failed to read from eventfd"), 1203);
+        let mut buf = [0u8; std::mem::size_of::<u64>()];
+        let _ = (&evt.0).read(buf.as_mut_slice());
```
you don't need this trick, just use `let mut evt = pipe()...` it's more clear

germag replied:
and belongs to the first commit

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -555,16 +504,18 @@ mod tests {
 
-        assert_eq!(evt.read().expect("failed to read from eventfd"), 1203);
+        let mut buf = [0u8; std::mem::size_of::<u64>()];
+        let _ = (&evt.0).read(buf.as_mut_slice());
```
you don't need this trick, just use `let mut evt = pipe()...`  it's more clear

germag replied:
also this belongs to the first commit

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -124,7 +70,7 @@ impl CmsgBuffer {
             (capacity.checked_add(size_of::<cmsghdr>()).unwrap() - 1) / size_of::<cmsghdr>();
         if capacity <= CMSG_BUFFER_INLINE_CAPACITY {
-            CmsgBuffer::Inline([0u64; (CMSG_BUFFER_INLINE_CAPACITY + 7) / 8])
+            CmsgBuffer::Inline([0u64; CMSG_BUFFER_INLINE_CAPACITY.div_ceil(8)])
```
as I mentioned, we should not add unrelated changes, I understand that clippy complains about this, but this code were already here. 

if you want to fix clippy warning it's ok, but should be done in a different PR, doing it here just makes the review more complex and can be confusing for the reviewers.

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -96,26 +56,12 @@ fn set_msg_controllen(msg: &mut msghdr, cmsg_capacity: usize) {
 
 enum CmsgBuffer {
-    Inline([u64; (CMSG_BUFFER_INLINE_CAPACITY + 7) / 8]),
+    Inline([u64; CMSG_BUFFER_INLINE_CAPACITY.div_ceil(8)]),
```
same, this belongs to a different PR

germag commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -27,7 +27,10 @@ fn new_msghdr(iovecs: &mut [iovec]) -> msghdr {
+        #[cfg(any(target_os = "linux", target_os = "android"))]
         msg_iovlen: iovecs.len(),
+        #[cfg(target_os = "macos")]
+        msg_iovlen: iovecs.len().try_into().unwrap(),
```
can at least use `.expect()`  with an explanation of what happened

germag commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -47,7 +50,14 @@ fn new_msghdr(iovecs: &mut [iovec]) -> msghdr {
+    #[cfg(target_os = "macos")]
+    {
+        msg.msg_controllen = cmsg_capacity.try_into().unwrap();
+    }
```
what about?
```rust
    msg.msg_controllen = if cfg!(target_os = "macos") {
        cmsg_capacity.try_into().expect("some descriptive error")
    } else {
        cmsg_capacity
    };
```

uran0sH replied:
It can't work. Because return type is different.

germag commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -202,7 +216,7 @@ unsafe fn raw_recvmsg(
                 // data must be dropped to insufficient buffer space for returning them to outer
                 // scope. This might be a sign of incorrect protocol communication.
-                for fd_offset in 0..fds_count {
+                for fd_offset in 0..fds_to_be_copied_count {
```
as I mentioned in my email, why this change?

uran0sH replied:
If use `fds_count`, the test will report an error on macos:
```
fatal runtime error: IO Safety violation: owned file descriptor already closed, aborting
```
Since the value of the `fds_count` is 4 on macos. But on linux, this value is 2. But my solution may be a wrong solution

germag replied:
ok we need to check this

germag replied:
but, this is the type of thing you must tell us (email, chat or in the meeting) instead of just making a change without any comment.

uran0sH replied:
Run send_more_recv_less2 or send_more_recv_less1
On macos, cmsg.cmsg_len is 28 and CMSG_LEN(0) is 12, so the fds_count is 4.
On linux, cmsg.cmsg_len is 24 and CMSG_LEN(0) is 16, the fds_count is 2.
Run send_recv_with_fd or send_recv_only_fd:
On macos, cmsg.cmsg_len is 16 and CMSG_LEN(0) is 12
On linux, cmsg.cmsg_len is 20 and CMSG_LEN(0) is 16

germag commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -481,6 +507,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg(any(target_os = "linux", target_os = "android"))]
```
why this test is disabled on macOS?

germag commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -98,7 +104,13 @@ impl CmsgBuffer {
+                .try_into()
+                .expect("size_of_val(out_fds) exceeds u32 range"),
+        ) as usize
+    };
```
This change belongs to commit 2. As I mentioned, instead of using `.unwrap()` use `.expect()`.

germag commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -113,8 +125,12 @@ fn raw_sendmsg<D: IntoIovec>(fd: RawFd, out_data: &[D], out_fds: &[RawFd]) -> Re
         let cmsg = cmsghdr {
-            // SAFETY: We ensure that the input value does not exceed the range of c_uint. c_uint to usize is a safe conversion
-            cmsg_len: unsafe { CMSG_LEN(size_of_val(out_fds).try_into().unwrap()) as usize },
+            #[cfg(any(target_os = "linux", target_os = "android"))]
+            // SAFETY: We ensure that the input value does not exceed the range of c_uint.
+            cmsg_len: unsafe { CMSG_LEN(size_of_val(out_fds).try_into().expect("size_of_value(out_fds) exceeds u32 range")) as usize },
+            #[cfg(target_os = "macos")]
+            // SAFETY: We ensure that the input value does not exceed the range of c_uint.
+            cmsg_len: unsafe { CMSG_LEN(size_of_val(out_fds).try_into().expect("size_of_value(out_fds) exceeds u32 range")) },
```
Maybe we can call CMSG_LEN.. once, assign that to a variable and just do the cast like:
```rust
let cmsg_len = unsafe { CMSG_LEN(size_of_val(out_fds).try_into().expect("size_of_value(out_fds) exceeds u32 range"))

...
cmsg_len:  if cfg!(target_os = "macos") { cmsg_len } else { cmsg_len as usize}
```

or something like this

uran0sH replied:
The statement if should return the same type. On macos, cmsg_len is u32, but cmsg_len as usize will return usize. So rust compiler will report an error here

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -151,7 +97,8 @@ impl CmsgBuffer {
 fn raw_sendmsg<D: IntoIovec>(fd: RawFd, out_data: &[D], out_fds: &[RawFd]) -> Result<usize> {
-    let cmsg_capacity = CMSG_SPACE!(std::mem::size_of_val(out_fds));
+    // SAFETY: CMSG_SPACE is a pure calculation. We ensure that the input value does not exceed the range of c_uint
+    let cmsg_capacity = unsafe { CMSG_SPACE(size_of_val(out_fds).try_into().unwrap()) as usize };
```
Replace `unwrap()` with `expect()`

germag commented on src/linux/sock_ctrl_msg.rs:
```diff
@@ -166,7 +113,8 @@ fn raw_sendmsg<D: IntoIovec>(fd: RawFd, out_data: &[D], out_fds: &[RawFd]) -> Re
         let cmsg = cmsghdr {
-            cmsg_len: CMSG_LEN!(std::mem::size_of_val(out_fds)),
+            // SAFETY: We ensure that the input value does not exceed the range of c_uint. c_uint to usize is a safe conversion
+            cmsg_len: unsafe { CMSG_LEN(size_of_val(out_fds).try_into().unwrap()) as usize },
```
Replace `unwrap()` with `expect()`

germag reviewed (commented):
just some small comments

stefano-garzarella commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -591,7 +619,6 @@ mod tests {
     #[test]
     fn send_more_recv_less2() {
         let (s1, s2) = UnixDatagram::pair().expect("failed to create socket pair");
-
```
unrelated change

stefano-garzarella commented on src/unix/sock_ctrl_msg.rs:
```diff
@@ -191,7 +207,7 @@ unsafe fn raw_recvmsg(
         if cmsg.cmsg_level == SOL_SOCKET && cmsg.cmsg_type == SCM_RIGHTS {
             let fds_count: usize =
-                ((cmsg.cmsg_len - CMSG_LEN(0) as usize) as usize) / size_of::<RawFd>();
+                ((cmsg.cmsg_len as usize - CMSG_LEN(0) as usize) as usize) / size_of::<RawFd>();
```
why you need this in this commit?

uran0sH replied:
cmsg.cmsg_len is u32 on macos

stefano-garzarella replied:
okay, I see, LGTM!

uran0sH committed:
Fix send_more_recv_less1/2() on macos

If we use fds_count, the test will report an error on macos:

```
fatal runtime error: IO Safety violation: owned file descriptor already closed, aborting
```

When running send_more_recv_less2 or send_more_recv_less1 on macos,
cmsg.cmsg_len is 28 and CMSG_LEN(0) is 12, which cause fds_count is 4.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

uran0sH committed:
Update CHANGELOG.md

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

dependabot[bot] committed:
build(deps): bump rust-vmm-ci from `c3ef897` to `aa2f8a9`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `c3ef897` to `aa2f8a9`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/c3ef8977388ccddb8c68b106737fa2c674b0d87a...aa2f8a90e46533de8a58f19cd0c417de856d4437)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-version: aa2f8a90e46533de8a58f19cd0c417de856d4437
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

roypat committed:
chore: appease clippy

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

uran0sH commented:
See: https://github.com/rust-vmm/vmm-sys-util/pull/245
